banyan/src/forest/read.rs

```diff
@@ -19,7 +13,295 @@ use anyhow::{anyhow, Result};
+                }
+
+                Ok(NodeInfo::Leaf(index, leaf)) => {
+                    let chunk = {
+                        if matches!(self.mode, Mode::Backward) {
+                            self.offset -= index.keys.count();
+                        }
+                        let mut matching: SmallVec<[_; 32]> = smallvec![true; index.keys.len()];
+                        self.query.containing(self.offset, index, &mut matching);
+                        let keys = index.select_keys(&matching);
+                        let elems: Vec<V> = match leaf.as_ref().select(&matching) {
+                            Ok(i) => i,
+                            Err(e) => return Some(Err(e)),
+                        };
+                        let offset = self.offset;
+                        let mut pairs = keys
+                            .zip(elems)
+                            .map(|((o, k), v)| (o + offset, k, v))
+                            .collect::<Vec<_>>();
+                        if matches!(self.mode, Mode::Backward) {
+                            pairs.reverse();
+                        }
+                        FilteredChunk {
+                            range: self.offset..self.offset + index.keys.count(),
+                            data: pairs,
+                            extra: (self.mk_extra)(IndexRef::Leaf(index)),
+                        }
+                    };
+                    if matches!(self.mode, Mode::Forward) {
+                        self.offset += index.keys.count();
+                    }
+
+                    // Ascend to parent's node
+                    self.stack.pop();
+                    let last = self.stack.last_mut().expect("not empty");
+                    last.next_pos(&self.mode);
+
+                    break chunk;
+                }
+
+                // even for purged leafs and branches or ignored chunks,
+                // produce a placeholder.
+                //
+                // the caller can find out if we skipped purged parts of the
+                // tree by using an appropriate mk_extra fn, or check
+                // `data.len()`.
+                Ok(_) => {
+                    let TraverseState { index, .. } = self.stack.pop().expect("not empty");
+                    // Ascend to parent's node. This might be none in case the
+                    // tree's root node is a `PurgedBranch`.
+                    if let Some(last) = self.stack.last_mut() {
+                        last.next_pos(&self.mode);
+                    }
+
+                    let range = match self.mode {
+                        Mode::Forward => {
+                            self.offset += index.count();
+                            self.offset - index.count()..self.offset
+                        }
+                        Mode::Backward => {
+                            self.offset -= index.count();
+                            self.offset + index.count()..self.offset
```

Comment:
This looks weird. So a range from x + n .. x, isn't that the empty range? I guess we don't have a test for this yet.

Also it would be nice if we could DRY the range creation.

Reply:
Why would it be the empty range? I flipped the bounds, as the data entries are also reversed, so that offset[0] corresponds to data[0]

Reply:
I tried DRYing up the range creation, into a `fn get_range(&mut self,..) but ran into lifetime (obviously) issues and ultimately decided it's not worth it. Not sure about another approach.

Reply:
https://doc.rust-lang.org/std/ops/struct.Range.html

"A (half-open) range bounded inclusively below and exclusively above (start..end).

The range start..end contains all values with start <= x < end. It is *empty* if start >= end."